### PR TITLE
Fix requester engine's optimization

### PR DESCRIPTION
**Pull request by zhangchiqing**
This PR fixes an issue in the requester engine that could cause EN not able to fetch collection.

### CODE REVIEW

File: engine/common/requester/engine.go [hunk 1]
```diff
@@ -346,31 +347,32 @@ func (e *Engine) dispatchRequest() (bool, error) {
...
-				filter.HasNodeID(providerID),
-				item.ExtraSelector,
-			))
-			if len(overlap) == 0 {
-				continue
-			}
-		}
+		// if providerID != flow.ZeroID {
+		// 	overlap := providers.Filter(filter.And(
+		// 		filter.HasNodeID(providerID),
```
Comment (zhangchiqing) [inline engine/common/requester/engine.go:352]: This optimization is supposed to fetch collections from the same collection node as much as possible. However, it's causing the collection for the unexecuted block to be "de-prioritized". 

After removing it. EN1 is able to continue executing blocks.

**Review by kc1116 — APPROVED** [review]

**Comment by janezpodhostnik** [comment]
Is this a temporary fix? will this have a big impact on performance overall?

File: engine/common/requester/engine.go (hunk 1 above)
Comment (zhangchiqing) [reply engine/common/requester/engine.go:352]: >I disagree that the purpose of the logic is to fetch from the same collection node as much as possible. Each individual entity request will be sent to a [randomly selected provider](https://github.com/onflow/flow-go/blob/b3eb73f6dee9851cc8d4b1938b664d1e76fd5248/engine/common/requester/engine.go#L370);

I don't think so.The dispatchRequest function is triggered either by a timer or by the `Force` function. Whenever it's triggered, it will only make one single request to a selected providerID with a list of entity IDs:
```
	err = e.con.Unicast(req, providerID)
```

The question is how is the providerID and the entity list selected? Since it only sends one requests, the optimization would like to bundle as many entities as possible to a single provider. We do randomly select a provider, but this provider is selected when processing the first entity, once a provider is selected, we will go through all the remaining entities and see if can be requested by the same provider. If not, it will skip to process the next. 

In other words, the dispatchRequest will send one request to the collection node which is determined by the first entity. 

Now the question is how is the first entity selected? It's random, since we shuffled the entities at the beginning of dispatchRequest.

The shuffled gives the provider of the next collection less chance to be selected. If the next collection is not fetched, then the next block won't be executed, and execution stops.

Comment (zhangchiqing) [reply engine/common/requester/engine.go:352]: The `48f626b8a62d6310ca5f41070e1ac3b716692adf39ae4030b6d73c1d32a57b33` collection is the next collection that EN1 is not fetching. 

I've been seeing this logs:
```
{"level":"error","node_role":"execution","node_id":"9686399a8a5418a12e762cfaeff2ea348c2137f554560917760e0d47acf2cda4","engine":"requester","error":"no valid providers available for item 48f626b8a62d6310ca5f41070e1ac3b716692adf39ae4030b6d73c1d32a57b33, total providers: 437","time":"2024-05-16T19:52:03.077986398Z","message":"could not dispatch requests"}
```

This means Line 365, the providers is empty:
```
providers = providers.Filter(item.ExtraSelector)
```

The overwriting of the providers variable seems problematic to me. Because if the first entity's providers are from cluster 2, then providers becomes cluster 2's collection nodes, then the next collection from cluster 1 will have no overlap with it, therefore will be skipped. So I commented out this section. The resulting behavior is that the request will send to the collection node determined by the first entity, which means, if the first collection determines to send request to a collection node from cluster 1, we also include the next collection in the batch request, even if it is guaranteed by cluster 2.

This explains your observation:

>This means we could include an entity in a request whose potential providers do not include providerID, but the provider might still (by chance) be able to fulfil the request anyway (it is part of the right cluster, but didn't happen to sign that collection). The majority of the time this wouldn't happen, though. This might be part of the reason why commenting this out is helping to fix the problem.

Comment (zhangchiqing) [reply engine/common/requester/engine.go:352]: Now the EN1 problem is solved by this change. The question is what would be the best strategy here?

To me, a good strategy must:
1. Give the next collection higher priority
This is important especially when falling far behind, in this case EN will fetch lots of collections, and if the next collection is not in higher priority, then the execution will stop.

2. Request as many as possible in a batch request

The above 2 is not enough, because if we always iterate from the same entity, and we never received this entity, then we would be always selecting a provider from the same cluster, which will give collections from other cluster less chance to be requested. I think that's why we introduced shuffling at the beginning. I hope we could get rid of this shuffle, can we? but at least, we can optimize the shuffling, it's currently O(n), where n is the number of pending items, I think we could do O(min(n, batchThreshold)), since we at most include batchThreshold items.

Comment (zhangchiqing) [reply engine/common/requester/engine.go:352]: Note when `no valid provider` error is hit, the dispatchRequest will not make any request, and then we will go into an endless loop, and not request any collection. So I think this is how my fix actually worked.

Comment (jordanschalm) [reply engine/common/requester/engine.go:352]: > I've been seeing this logs:
```... no valid providers available for item ...```
This means Line 365, the providers is empty:
```providers = providers.Filter(item.ExtraSelector)```

To me this is the core unexpected behaviour here and what we need to understand.
- We set `providers` initially to `state.Final().Identities()` (line 310)
- We randomly select one item from `items`
  - We attempt to find the intersection between the potential providers reported by the item (`ExtraSelector`) and all nodes in the network, and **fail to find any, which should never happen ⚠️**. 

The fact that `providers.Filter(item.ExtraSelector)` returns an empty list indicates either that the item is specifying `ExtraSelector` incorrectly/incompletely or there is some severe unexpected discrepancy in the identity table state.

### Why would `ExtraSelector` have no overlap with `state.Final().Identities()`?

Originally `ExtraSelector` simply matched all the collection guarantors (~2/3 of the cluster). Under these conditions, `ExtraSelector` should always match a non-zero number of providers from `state.Final().Identities()`.

However I noticed we added an option called `onflowOnlyLNs`, which will modify `ExtraSelector` to select nodes which BOTH:
- were guarantors of the collection (~2/3s of cluster)
- are operated by Flow Foundation, by regex (~40-50% of collection nodes)

https://github.com/onflow/flow-go/blob/150dfeed88629189265d30324b6320d89819acfc/engine/execution/ingestion/fetcher/fetcher.go#L58-L81

With these conditionals together, I think we can arrive at a case where `ExtraSelector` matches no providers.

- There are 106 Collection nodes currently, so clusters of size 35-36.
- Last epoch (where this problem occurred) we had an unusually uneven distribution for the `.*onflow.org.*` nodes, with 11 in one cluster and 18-19 in the other two.
- with cluster size 35, `f=11` and we need `2f+1=23` signatures
  - The number of cluster members who don't size is `35-23=12 > 11`.
  - **So we can have a collection which has 0 `onflow.org` signers.**
image

Do you know if that option was enabled at the time? 

## Other comments

> The overwriting of the providers variable seems problematic to me.

I don't follow the reasoning here. We don't overwrite `providers` at all in a meaningful way. The variable is written to twice, on line 310 and 375, but you could remove the write on line 310 without changing the behaviour (just replace `providers` with `state.Final().Identities()` on line 366). 

Functionally, `providers` is written at most once per `dispatchRequest` call, on line 375.

> Give the next collection higher priority

It's worth noting that `items` is a map. I don't think we have any notion of collection ordering within the `requester` engine. 

> Request as many as possible in a batch request

👍 Agree. As I mentioned in my comment, I think the code currently commented out, which may omit items from a batch request, is necessary to avoid sending many requests to providers who cannot fulfil them. 

I think a fairly simple optimization would be to:
- select `providerID` immediately before sending the request, rather than after processing the first `item`
- compute the overlap as `filter.And(filter.HasNodeID(providers.NodeIDs()...), item.ExtraSelector)`

This way, we are more likely to include more items in each request.

> Note when no valid provider error is hit, the dispatchRequest will not make any request, and then we will go into an endless loop, and not request any collection

This is a fairly severe unexpected failure in my opinion. I think the current behaviour of exiting and logging an error (with more scary error/log messages making it clear this is not expected) is reasonable. 

If it's true that this error prevents _any_ collections from being requested, I agree the behaviour should be changed. However, my interpretation (based on the randomization of `items`) is that it only prevents the problematic collection from being requested. Subsequent `dispatchRequests` calls, so long as they select a different item first, will be able to request collections as expected.

Comment (zhangchiqing) [reply engine/common/requester/engine.go:352]: >With these conditionals together, I think we can arrive at a case where ExtraSelector matches no providers.

Yes, good point. It's indeed the case. I did captured the following log which logs the list of guarantees for the missing collection:

```
{"level":"info","node_role":"execution","node_id":"9686399a8a5418a12e762cfaeff2ea348c2137f554560917760e0d47acf2cda4","component":"ingestion_engine_collection_fetcher","onflowOnlyLNs":true,"time":"2024-05-16T22:49:18.243538517Z","message":"queued collection 48f626b8a62d6310ca5f41070e1ac3b716692adf39ae4030b6d73c1d32a57b33 for block 135f860835d05303575d24bea5c90e06182b779cf49c05a958570f5b737e7237, height 77973667 from guarantors: [050dfd9898470c042e9190b8a9136122dee26dad438cf5b49307f55fe4dbcc98 095c8178b53488ab9096eb05c2d8d0545a6a492e81449d36d32f6ac73e052a94 119984b82815a2093211f50e55a5bb00c2915c86b13ebab97a4019865149a880 1408c08272beb24ad4fe4e531a9b937a26ee72c5f56b50ac8f2b889c3c42d316 1704a8abb18a5e7252b61442da19e6acfc810b8fead4d3e7a69339082c8195e4 1eb1354ea8e4511dfd5686e5e5e01a328acb2904de9a8f7f0f0408cdde1a5705 21cccaf4414a774633d27a8387dad4acf7799b0c15816e46780f2fae3fc115f3 40344416089a04c0574fded8db70c310dc105edf4dcee4651b10dd5205b966cc 419b51adb754493e2beb96b87dba20ad9f5a67fe6f0f63c54a7e990791407aeb 48ba5aabeaf1a7c1bc6a96b20c8e9f6cbe7f6b3c77faec3b77fa66a630abf8c7 59d9960277d26d9c4ee50968e741eb60a0a63663bcbbcd6ab9e4b176bc2ba19b 5cda551830e62c9f7673fc22e5853ab07f00441e9757ace2360d340d523c86ae 67f197c51081839eeb4e119c47f17cadce98544958fd00aaefa9e252b299d0da 6c8c3eaa2d0476d3d4144ca9960654b3bf36293f4d1d3cf0e69f04269e7dad7e 9a6d6a6adfb99d96c75f372b71bf8d041b5bf35cba18a1ad25b3a04a45365dfe bd2822c1e656b30701b1fd0910ec004c63b56744b41cb4bc489ddee7642b0009 cb45f5e3592ee68612fee0a8b69e3f2b3fa3c5ca45b1af2b6895e8f61a43487f d310ae34543c906c005c0ccbb19f4b38258650144c427ccd0e796b777a1e700a de84f43f4282ee8be636b46c67603a732d337324ebcd02885b7bc871042920ed e20501aa4f972ddeddbd3905566389109b02b8eacd2d0dc48a4f2a567dedd648 e7d97464e5fda01b49015a8a795766e97382287b87fb3582a65d07598f0d8f69 f09574d4fba0900e8be8b0ec0e5bf28c5aad0a1ca4b54323dce3ce0b41fd968b f6d8506ee610241aed91e8b31c49bc4cc992474bae3e360bafdc066adca3c005 fe14c6a722212e19af1c143008d46413d4cea5097714a3720c67f356fb7c2982]"}
```

Non of the node IDs are from onflow.org, even through some of them are from dapper-flow.com. 

And EN1 does have the flag on which makes `onflowOnlyLNs=true`, and this caused `no valid provider` error. 

In terms of solution, I think we can maybe simply extend the onflowOnLNs to onflow.org and dapper-flow.com, and keep the rest logic as is.

File: engine/common/requester/engine.go [hunk 2]
```diff
@@ -306,29 +306,17 @@ func (e *Engine) dispatchRequest() (bool, error) {
...
 	// get the current top-level set of valid providers
-	providers, err := e.state.Final().Identities(e.selector)
+	final := e.state.Final()
+	providers, err := final.Identities(e.selector)
 	if err != nil {
 		return false, fmt.Errorf("could not get providers: %w", err)
 	}
 
-	// randomize order of items, so that they can be requested in different order each time
-	rndItems := make([]flow.Identifier, 0, len(e.items))
```
Comment (zhangchiqing) [inline engine/common/requester/engine.go:315]: The e.items is a map, iterating it is already close to randomly iterating. We don't need to explicitly shuffle the items here. And in fact, doing this would increase time complexity to O(len(e.items)). Removing this random shuffle, could reduce the time complexity to O(max(BatchThreshold, len(e.items))).

File: engine/execution/ingestion/fetcher/fetcher.go [hunk 3]
```diff
@@ -12,7 +12,7 @@ import (
 	"github.com/onflow/flow-go/state/protocol"
 )
 
-var onlyOnflowRegex = regexp.MustCompile(`.*\.onflow\.org:3569$`)
+var onlyOnflowRegex = regexp.MustCompile(`.*\.(onflow\.org|dapper-flow\.com):3569$`)
```
Comment (zhangchiqing) [inline engine/execution/ingestion/fetcher/fetcher.go:15]: Expanding the whitelist here so that it's guaranteed to have at least one collection node available to fetch collection from.

File: engine/common/requester/engine.go (hunk 1 above)
Comment (zhangchiqing) [reply engine/common/requester/engine.go]: After discussion, we learned the reason commenting out the code that skip the collection if the randomly selected provider can not provider the collection is that, without that logic, it will eventually hit a collection where one of its guarantors belongs to the same cluster of the missing collection, but not the guarantors of the missing collection. So it forced the requester to request that collection from a collection node of the same cluster but didn't sign the collection, which could also work.

**Review by jordanschalm — APPROVED** [review]
